File: packages/packagechk/src/packagechk/output.py
from __future__ import annotations

import json

from packagechk.models import CheckStatus, PackageCheckReport, RegistryCheckResult


def render_json(report: PackageCheckReport) -> str:
    return json.dumps(report.to_json_dict(), sort_keys=True)


def render_human(report: PackageCheckReport) -> str:
    return "\n".join(_render_result(result) for result in report.results)


def _render_result(result: RegistryCheckResult) -> str:
    lookup_suffix = ""
    if result.lookup_name != result.input_name:
        lookup_suffix = f" as {result.lookup_name!r}"

    if result.status is CheckStatus.AVAILABLE:
        return f"{result.registry.value}: available{lookup_suffix}"
    if result.status is CheckStatus.TAKEN:
        details = [f"{result.registry.value}: taken{lookup_suffix}"]
        if result.latest_version is not None:
            details.append(f"latest {result.latest_version}")
        if result.description is not None:
            details.append(result.description)
        if result.package_url is not None:
            details.append(result.package_url)
        return " — ".join(details)
    return f"{result.registry.value}: {result.status.value}: {result.message}"

File: ts/packages/packagechk/src/output.ts
import {
	reportToJson,
	type CheckStatus,
	type PackageCheckReport,
	type Registry,
	type RegistryCheckResult,
} from "./models.ts";

export function renderJson(report: PackageCheckReport): string {
	return JSON.stringify(sortJsonValue(reportToJson(report)));
}

export function renderHuman(report: PackageCheckReport): string {
	return report.results.map(renderResult).join("\n");
}

function renderResult(result: RegistryCheckResult): string {
	const lookupSuffix =
		result.lookupName === result.inputName
			? ""
			: ` as ${JSON.stringify(result.lookupName).replaceAll('"', "'")}`;
	if (result.status === "available") return `${result.registry}: available${lookupSuffix}`;
	if (result.status === "taken") {
		const details = [`${result.registry}: taken${lookupSuffix}`];
		if (result.latestVersion !== undefined) details.push(`latest ${result.latestVersion}`);
		if (result.description !== undefined) details.push(result.description);
		if (result.packageUrl !== undefined) details.push(result.packageUrl);
		return details.join(" — ");
	}
	return formatRegistryStatusLine(result.registry, result.status, result.message);
}

export function formatRegistryStatusLine(
	registry: Registry,
	status: CheckStatus,
	message: string,
): string {
	return `${registry}: ${status}: ${message}`;
}

function sortJsonValue(value: unknown): unknown {
	if (Array.isArray(value)) return value.map(sortJsonValue);
	if (value === null || typeof value !== "object") return value;
	const entries = Object.entries(value as Record<string, unknown>).sort(([left], [right]) =>
		left < right ? -1 : left > right ? 1 : 0,
	);
	return Object.fromEntries(entries.map(([key, child]) => [key, sortJsonValue(child)]));
}
